pandas/_libs/index.pyx

```diff
@@ -649,7 +649,89 @@ cdef class BaseMultiIndexCodesEngine:
     def _codes_to_ints(self, codes):
         raise NotImplementedError("Implemented by subclass")
 
-    def _extract_level_codes(self, object target):
+    def _do_backfill_carrying(self, object level_codes,
+                              object level_codes_no_fill):
+        """
+        cleanup for the case of per-level backfills of too-large values to -1
+
+        given a 2d list of level_codes, i.e. integers representing the index
+        of a target index inside `self.level_codes` for each level, which
+        were backfilled at each level, handle cases where too-large values
+        which were backfilled to 0 (and became 1 after adding 1 to handle use
+        of unsigned integers) and thus the backfilling on the level code, when
+        represented as an int, will not be correct
+
+        in particular, this involves a sort of "carrying", whereby, when a
+        value at a given level is too large, we set it to the minimum value
+        and then attempt to bump the value to the left, and, in the event that
+        we need to do this at the first index
+
+        e.g. if the highest two values in `self`'s index tuples are (8, 4, 7)
+        and (8, 4, 9), then, in order to determine that the value (8, 4, 10) is
+        too large, we would need to determine that 10 > 9, meaning we need to
+        "carry" to the middle level, at which we try to bump 4 to 5, which is
+        in turn too large, requiring us to try to bump 8 to 9, also too large,
+        meaning that the entire row representing (8, 4, 10)'s level_codes
+        should be set to -1
+
+        Parameters
+        ----------
+        level_codes : ndarray[ndim=2]
```

Comment:
do we have a dtype here?  or maybe its int-like but we dont know which intlike?

Reply:
good point -- this method should work with any int-like type but its only use is with an array with `dtype='uint64'` (see [here](https://github.com/pandas-dev/pandas/pull/30766/files#diff-495f440aad68721ad5ca1f35087fc450R783), specifically).

what's the preferred way to represent a 2d np array of int-like values in a docstring here?

Reply:
`ndarray[uint64_t, ndim=2]`, which you might be able to add as a type in the signature